### PR TITLE
Add polling for tests and fix rarely breaking tests

### PR DESCRIPTION
**Pull request by i1i1**
https://jira.hyperledger.org/browse/IR-1008
Signed-off-by: i1i1 <vanyarybin1@live.ru>

### Description of the Change

This pr reworks tests with simple test library. Also removes timeouts and adds polling with timeout.

### Benefits

Tests won't unconditionally fail because of weird environment.

### Possible Drawbacks 

Sometimes tests can be slower, but at least they won't signal false positive breaking of iroha.

### CODE REVIEW

**Review by rkharisov — APPROVED** [review]

**Review by rkharisov — COMMENTED** [review]
Good job!

**Comment by eadventurous** [comment]
A general comment, I don't think it should be named `Fixup breaking tests`. This not a simple fix for tests, better something like `Adds polling to tests`

File: iroha_client/tests/tests/tests.rs [hunk 1]
```diff
@@ -13,3 +29,271 @@ mod transfer_asset;
...
+                    return Some(asset.as_ref());
+                }
+            }
+            None
+        })
+    }
+}
+
+#[allow(clippy::clippy::cast_possible_truncation)]
+impl TestNetwork for Network {
```
Comment (eadventurous) [inline iroha_client/tests/tests/tests.rs:254]: From what I remember `Network` here is already a `test_network`? Should we simply move all this functions inside of it without the trait?

File: iroha_client/tests/tests/tests.rs [hunk 2]
```diff
@@ -13,3 +29,271 @@ mod transfer_asset;
...
+    fn start_test(n_peers: usize, maximum_transactions_in_block: u32) -> (Self, Client) {
+        let mut configuration = Configuration::test();
+        // Reverse consensus
+        configuration.sumeragi_configuration.max_faulty_peers = ((n_peers - 1) / 3) as u32;
```
Comment (eadventurous) [inline iroha_client/tests/tests/tests.rs:258]: What do you mean by `Reverse consensus`?

File: iroha_client/tests/tests/tests.rs [hunk 3]
```diff
@@ -13,3 +29,271 @@ mod transfer_asset;
...
+        configuration
+            .queue_configuration
+            .maximum_transactions_in_block = maximum_transactions_in_block;
+        let network = Network::new(Some(configuration), n_peers).expect("Failed to init peers");
```
Comment (eadventurous) [inline iroha_client/tests/tests/tests.rs:262]:
```suggestion
        let network = Network::new(Some(configuration), n_peers).expect("Failed to initialize peers.");
```

File: iroha_client/src/client.rs [hunk 4]
```diff
@@ -80,8 +80,8 @@ impl Client {
     /// # Errors
     /// Fails if sending transaction to peer fails or if it response with error
     #[log]
-    pub fn submit(&mut self, instruction: Instruction) -> Result<Hash> {
-        self.submit_all(vec![instruction])
+    pub fn submit(&mut self, instruction: impl Into<Instruction> + Debug) -> Result<Hash> {
```
[the remark this answers is not in the document]
Comment (i1i1) [reply iroha_client/src/client.rs:83]: Yes, if want to `#[log]` it

File: iroha_client/tests/tests/tests.rs (hunk 1 above)
Comment (i1i1) [reply iroha_client/tests/tests/tests.rs:254]: As for me it seems like a bunch of dirty hacks which makes our tests running. That crate only contains logic for spawning bunch of peers.

File: iroha_client/tests/tests/tests.rs (hunk 2 above)
Comment (i1i1) [reply iroha_client/tests/tests/tests.rs:258]: I mean that we can calculate how many peers can be faulty from consensus requirements

Comment (eadventurous) [reply iroha_client/tests/tests/tests.rs:258]: Better to write it then like `Calculates max faulty peers based on consensus requirements`.

**Review by eadventurous — APPROVED** [review]

**Pull request merged**